File: android/app/src/main/java/com/molt/marketplace/core/designsystem/component/MoltPriceText.kt
package com.molt.marketplace.core.designsystem.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.tooling.preview.Preview
import com.molt.marketplace.R
import com.molt.marketplace.core.designsystem.theme.MoltColors
import com.molt.marketplace.core.designsystem.theme.MoltSpacing
import com.molt.marketplace.core.designsystem.theme.MoltTheme

@Composable
fun MoltPriceText(
    price: String,
    modifier: Modifier = Modifier,
    originalPrice: String? = null,
    currencySymbol: String = "EUR",
    size: MoltPriceSize = MoltPriceSize.Medium,
) {
    val priceStyle: TextStyle
    val originalPriceStyle: TextStyle

    when (size) {
        MoltPriceSize.Small -> {
            priceStyle = MaterialTheme.typography.bodySmall
            originalPriceStyle = MaterialTheme.typography.labelSmall
        }
        MoltPriceSize.Medium -> {
            priceStyle = MaterialTheme.typography.titleMedium
            originalPriceStyle = MaterialTheme.typography.bodySmall
        }
        MoltPriceSize.Large -> {
            priceStyle = MaterialTheme.typography.headlineSmall
            originalPriceStyle = MaterialTheme.typography.bodyMedium
        }
    }

    val formattedPrice = "$currencySymbol $price"
    val formattedOriginalPrice = originalPrice?.let { "$currencySymbol $it" }

    val accessibilityDescription = if (formattedOriginalPrice != null) {
        stringResource(R.string.common_sale_price_label, formattedPrice, formattedOriginalPrice)
    } else {
        stringResource(R.string.common_price_label, formattedPrice)
    }

    Row(
        modifier = modifier.semantics { contentDescription = accessibilityDescription },
        horizontalArrangement = Arrangement.spacedBy(MoltSpacing.SM),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = formattedPrice,
            style = priceStyle,
            color = if (originalPrice != null) MoltColors.PriceSale else MoltColors.PriceRegular,
        )

        if (formattedOriginalPrice != null) {
            Text(
                text = formattedOriginalPrice,
                style = originalPriceStyle.copy(textDecoration = TextDecoration.LineThrough),
                color = MoltColors.PriceOriginal,
            )
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun MoltPriceTextRegularPreview() {
    MoltTheme {
        MoltPriceText(price = "29.99")
    }
}

@Preview(showBackground = true)
@Composable
private fun MoltPriceTextSalePreview() {
    MoltTheme {
        MoltPriceText(price = "29.99", originalPrice = "39.99")
    }
}

@Preview(showBackground = true)
@Composable
private fun MoltPriceTextLargePreview() {
    MoltTheme {
        MoltPriceText(price = "199.99", originalPrice = "249.99", size = MoltPriceSize.Large)
    }
}

@Preview(showBackground = true)
@Composable
private fun MoltPriceTextSmallPreview() {
    MoltTheme {
        MoltPriceText(price = "9.99", size = MoltPriceSize.Small)
    }
}

File: ios/XiriGoEcommerce/Core/DesignSystem/Component/MoltPriceText.swift
import SwiftUI

// MARK: - MoltPriceSize

enum MoltPriceSize {
    case small
    case medium
    case large

    var priceFont: Font {
        switch self {
        case .small: return MoltTypography.bodySmall
        case .medium: return MoltTypography.titleMedium
        case .large: return MoltTypography.headlineSmall
        }
    }

    var originalPriceFont: Font {
        switch self {
        case .small: return MoltTypography.labelSmall
        case .medium: return MoltTypography.bodySmall
        case .large: return MoltTypography.bodyMedium
        }
    }
}

// MARK: - MoltPriceText

struct MoltPriceText: View {
    // MARK: - Properties

    private let price: String
    private let originalPrice: String?
    private let currencySymbol: String
    private let size: MoltPriceSize

    // MARK: - Init

    init(
        price: String,
        originalPrice: String? = nil,
        currencySymbol: String = "EUR",
        size: MoltPriceSize = .medium
    ) {
        self.price = price
        self.originalPrice = originalPrice
        self.currencySymbol = currencySymbol
        self.size = size
    }

    // MARK: - Body

    var body: some View {
        HStack(spacing: MoltSpacing.sm) {
            Text("\(currencySymbol) \(price)")
                .font(size.priceFont)
                .foregroundStyle(hasSale ? MoltColors.priceSale : MoltColors.priceRegular)

            if let originalPrice {
                Text("\(currencySymbol) \(originalPrice)")
                    .font(size.originalPriceFont)
                    .foregroundStyle(MoltColors.priceOriginal)
                    .strikethrough()
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityDescription)
    }

    // MARK: - Private

    private var hasSale: Bool {
        originalPrice != nil
    }

    private var accessibilityDescription: String {
        if let originalPrice {
            return String(
                localized: "common_sale_price_label \(currencySymbol) \(price) \(currencySymbol) \(originalPrice)"
            )
        }
        return String(
            localized: "common_price_label \(currencySymbol) \(price)"
        )
    }
}

// MARK: - Previews

#Preview("MoltPriceText Regular") {
    VStack(alignment: .leading, spacing: MoltSpacing.sm) {
        MoltPriceText(price: "29.99", size: .small)
        MoltPriceText(price: "29.99", size: .medium)
        MoltPriceText(price: "29.99", size: .large)
    }
    .padding()
}

#Preview("MoltPriceText Sale") {
    VStack(alignment: .leading, spacing: MoltSpacing.sm) {
        MoltPriceText(price: "19.99", originalPrice: "29.99", size: .small)
        MoltPriceText(price: "19.99", originalPrice: "29.99", size: .medium)
        MoltPriceText(price: "19.99", originalPrice: "29.99", size: .large)
    }
    .padding()
}
